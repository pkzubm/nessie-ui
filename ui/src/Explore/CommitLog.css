/**
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.commitLog__messageLink {
    color: #000 !important;
    font-weight: 600;
}
.commitLog__messageLink:hover {
    color: #007bff !important;
    text-decoration: underline;
}
.commitLog__body{
    display: flex;
    justify-content: space-between;
    align-items: center;
}
.commitLog__btnWrapper{
    display: flex;
    align-items: center;
}
.commitLog__hashBtnWrapper{
    display: flex;
    align-items: center;
    background-color: #f5f7f9;
}
.commitLog__notFoundText{
    display: flex;
    justify-content: center;
}
.commitLog__copyBtn{
    color: #000 !important;
}
.commitLog__hashBtn{
    color: #0969da !important;
}
.rightBtnHover:hover{
    color: #f5f7f9 !important;
    background-color: #0969da;
}
.commitLog__pagination{
    padding: 10px;
    justify-content: flex-end;
}