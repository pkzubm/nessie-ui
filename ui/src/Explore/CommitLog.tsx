/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Button, Card, Nav } from "react-bootstrap";
import moment from "moment";
import React, { useEffect, useState, Fragment } from "react";
import { useParams, useHistory, useLocation } from "react-router-dom";
import { api, CommitMeta } from "../utils";
import { factory } from "../ConfigLog4j";
import CodeIcon from "@material-ui/icons/Code";
import "./CommitLog.css";
import { Icon, Tooltip } from "@material-ui/core";
import ExploreLink from "./ExploreLink";
import { routeSlugs } from "./Constants";
import { EmptyMessageView } from "./Components";
import CommitDetails from "./CommitDetails";

const log = factory.getLogger("api.CommitHeader");

const fetchLog = (ref: string): Promise<void | CommitMeta[] | undefined> => {
  return api()
    .getCommitLog({ ref })
    .then((data) => {
      if (data.operations && data.operations.length > 0) {
        return data.operations;
      }
    })
    .catch((t) => log.error("CommitLog", t));
};

const CommitLog = (props: {
  currentRef: string;
  path: string[];
}): React.ReactElement => {
  const { currentRef, path } = props;
  const { slug } = useParams<{ slug: string }>();
  const [logList, setLogList] = useState<CommitMeta[]>([
    {
      author: undefined,
      authorTime: undefined,
      commitTime: undefined,
      committer: undefined,
      hash: undefined,
      message: "",
      properties: {},
      signedOffBy: undefined,
    },
  ]);
  const [showCommitDetails, setShowCommitDetails] = useState(false);
  const [commitDetails, setCommitDetails] = useState<CommitMeta | undefined>();
  const location = useLocation();
  const history = useHistory();
  useEffect(() => {
    const logs = async () => {
      const results = await fetchLog(currentRef);
      if (results) {
        setLogList(results);
      }
      if (showCommitDetails) {
        const listPath = location.pathname.substring(
          0,
          location.pathname.lastIndexOf("/")
        );
        history.push(listPath);
      }
    };
    void logs();
  }, [currentRef]);

  useEffect(() => {
    if (slug) {
      const last = slug.substring(slug.lastIndexOf("/") + 1, slug.length);
      setShowCommitDetails(last !== routeSlugs.commits);
      const logDetails = logList.find((logItem) => logItem.hash === last);
      setCommitDetails(logDetails);
    }
  }, [slug, logList]);

  if (!logList || (logList.length === 1 && !logList[0].hash)) {
    return <EmptyMessageView />;
  }

  const copyHash = async (hashCode: string) => {
    await navigator.clipboard.writeText(hashCode);
  };
  const commitList = (currentLog: CommitMeta, index: number) => {
    const { commitTime, author, message, hash } = currentLog;
    const hoursDiff = moment().diff(moment(commitTime), "hours");
    const dateTimeAgo =
      hoursDiff > 24
        ? moment(commitTime).format("YYYY-MM-DD, hh:mm a")
        : `${moment(commitTime).fromNow()}`;

    return (
      <Fragment key={index}>
        <Card.Body className="commitLog__body border-bottom">
          <div>
            <Nav.Item>
              <ExploreLink
                toRef={currentRef}
                path={path.concat(hash ?? "#")}
                type="CONTAINER"
                className="commitLog__messageLink"
              >
                {message}
              </ExploreLink>
            </Nav.Item>
            <Card.Text className={"ml-3"}>
              {author}
              <span className={"ml-2"}>committed on</span>
              <span className={"ml-2"}>{dateTimeAgo}</span>
            </Card.Text>
          </div>
          <div className={"commitLog__btnWrapper"}>
            <div className={"border rounded commitLog__hashBtnWrapper"}>
              <Tooltip title="Copy hash">
                <Button
                  className="border-right commitLog__copyBtn rightBtnHover"
                  variant="link"
                  onClick={() => copyHash(hash || "")}
                >
                  <Icon>content_copy</Icon>
                </Button>
              </Tooltip>
              <ExploreLink
                toRef={currentRef}
                path={path.concat(hash ?? "#")}
                type="CONTAINER"
                className="commitLog__hashBtn rightBtnHover"
              >
                <Tooltip title="Commit details">
                  <Button variant="link">
                    <span className="font-italic">{hash?.slice(0, 8)}</span>
                  </Button>
                </Tooltip>
              </ExploreLink>
            </div>
            <div>
              <Tooltip title="Browse the repository at this point in the history">
                <Button variant="light" className={"ml-3 rightBtnHover"}>
                  <CodeIcon />
                </Button>
              </Tooltip>
            </div>
          </div>
        </Card.Body>
      </Fragment>
    );
  };

  return (
    <Card className={"commitLog"}>
      {!showCommitDetails ? (
        logList.map((item, index) => {
          return commitList(item, index);
        })
      ) : (
        <CommitDetails commitDetails={commitDetails} currentRef={currentRef} />
      )}
    </Card>
  );
};

export default CommitLog;
