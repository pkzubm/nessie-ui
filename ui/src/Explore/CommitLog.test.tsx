/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import { BrowserRouter } from "react-router-dom";
import { render, screen, waitFor } from "@testing-library/react";
import CommitLog from "./CommitLog";
// tslint:disable-next-line:no-implicit-dependencies
import nock from "nock";

it("Commit log renders", async () => {
  const now = new Date();
  now.setDate(now.getDate() - 1);
  const commitMeta = {
    hash: "deadbeef",
    author: "bob",
    commitTime: now,
    committer: "sally",
    message: "commitMessage",
    properties: { a: "b", c: "d" },
  };
  const scope1 = nock("http://localhost/api/v1")
    .get("/trees/tree/main/log")
    .reply(200, { token: "foo", operations: [commitMeta] });

  const { getByText, asFragment } = render(
    <React.StrictMode>
      <BrowserRouter>
        <CommitLog currentRef={"main"} path={["main"]} />
      </BrowserRouter>
    </React.StrictMode>
  );
  expect(asFragment()).toMatchSnapshot();
  await waitFor(() => getByText("deadbeef"));

  scope1.done();
  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByText("deadbeef")).toBeInTheDocument();
});
