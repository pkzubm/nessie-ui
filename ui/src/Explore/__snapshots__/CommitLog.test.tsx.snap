// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commit log renders 1`] = `
<DocumentFragment>
  <div
    class="EmptyView card-header"
  >
    <span
      class="EmptyView__message"
    >
      Nothing to show
    </span>
  </div>
</DocumentFragment>
`;

exports[`Commit log renders 2`] = `
<DocumentFragment>
  <div
    class="commitLog card"
  >
    <div
      class="commitLog__body border-bottom card-body"
    >
      <div>
        <div
          class="nav-item"
        >
          <a
            class="commitLog__messageLink"
            href="/tree/main/main/deadbeef"
          >
            commitMessage
          </a>
        </div>
        <p
          class="ml-3 card-text"
        >
          bob
          <span
            class="ml-2"
          >
            committed on
          </span>
          <span
            class="ml-2"
          >
            a day ago
          </span>
        </p>
      </div>
      <div
        class="commitLog__btnWrapper"
      >
        <div
          class="border rounded commitLog__hashBtnWrapper"
        >
          <button
            class="border-right commitLog__copyBtn rightBtnHover btn btn-link"
            title="Copy hash"
            type="button"
          >
            <span
              aria-hidden="true"
              class="material-icons MuiIcon-root"
            >
              content_copy
            </span>
          </button>
          <a
            class="commitLog__hashBtn rightBtnHover"
            href="/tree/main/main/deadbeef"
          >
            <button
              class="btn btn-link"
              title="Commit details"
              type="button"
            >
              <span
                class="font-italic"
              >
                deadbeef
              </span>
            </button>
          </a>
        </div>
        <div>
          <button
            class="ml-3 rightBtnHover btn btn-light"
            title="Browse the repository at this point in the history"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Commit redirects on an invalid ref 1`] = `
<DocumentFragment>
  <div
    class="EmptyView card-header"
  >
    <span
      class="EmptyView__message"
    >
      Nothing to show
    </span>
  </div>
</DocumentFragment>
`;
